http4k-core/src/main/kotlin/org/http4k/filter/ResponseFilters.kt

```diff
@@ -131,6 +135,74 @@ object ResponseFilters {
     fun Base64EncodeBody() = Filter { next ->
         { next(it).run { body(Base64.getEncoder().encodeToString(body.payload.array())) } }
     }
+
+    class EtagSupport : Filter {
+        private val md5 = ThreadLocal.withInitial {
+            try {
+                MessageDigest.getInstance("MD5")
+            } catch (e: NoSuchAlgorithmException) {
+                throw RuntimeException("Unable to instantiate MD5 digest", e)
+            }
+        }
+
+        override fun invoke(next: HttpHandler): HttpHandler = { request ->
+            next(request).let { response ->
+                if (request.method == Method.GET && response.status == Status.OK) {
+                    checkEtag(request, response)
+                } else response
+            }
+        }
+
+        private fun checkEtag(request: Request, response: Response): Response {
+            val upstreamEtag = response.header("etag")
+            return if (upstreamEtag == null && isStreamingResponse(response)) {
+                response
+            } else {
+                val etag = upstreamEtag ?: generateStrongEtag(response.body.payload.array())
+                when {
+                    etag == request.header("if-none-match") -> {
+                        closeOldResponse(response)
+                        Response(Status.NOT_MODIFIED).withSafeHeadersFrom(response)
+                    }
+                    upstreamEtag == null -> response.header("etag", etag)
+                    else -> response
+                }
+            }
+        }
+
+        private fun closeOldResponse(response: Response) =
+            try {
+                response.close()
+            } catch (ignored: Exception) {
+            }
+
+        private fun isStreamingResponse(response: Response): Boolean = response.body.length == null
+
+        private fun Response.withSafeHeadersFrom(sourceResponse: Response) =
+            SAFE_HEADERS.fold(this) { response, headerName ->
+                response.headerIfPresent(
+                    headerName,
+                    sourceResponse.header(headerName)
+                )
+            }
+
+        private fun Response.headerIfPresent(name: String, value: String?) = if (value != null) {
+            this.header(name, value)
+        } else {
+            this
+        }
+
+        private fun generateStrongEtag(content: ByteArray): String = "\"${hash(content).toHex()}\""
+
+        private fun hash(bytes: ByteArray): ByteArray = md5.get().digest(bytes)
```

Comment:
why here you are using regular function and below extension? maybe this could be unified?

Reply:
As said, this is a port from our internal code, which was original ported from Java and Utterlyidle, and has been modified by multiple people over the years. Upside, it's prod tested. Downside, there are inconsistencies. Free free to clean it up, it could probably do with it.

Reply:
Done: https://github.com/http4k/http4k/commit/0365ad2a11a443f4a665c2dc871b62e98dedf7f0